Assets/MixedRealityToolkit-SDK/Features/Audio/Influencers/AudioInfluencerController.cs

```diff
@@ -0,0 +1,297 @@
+        [SerializeField]
+        private int MaxObjects = 10;
+
+        /// <summary>
+        /// Time of last audio processing update. 
+        /// </summary>
+        private DateTime lastUpdate = DateTime.MinValue;
+
+        /// <summary>
+        /// The source of the audio.
+        /// </summary>
+        private AudioSource audioSource;
+
+        /// <summary>
+        /// The initial volume level of the audio source.
+        /// </summary>
+        private float initialAudioSourceVolume;
+
+        /// <summary>
+        /// The hits returned by Physics.RaycastAll
+        /// </summary>
+        private RaycastHit[] hits;
+
+        /// <summary>
+        /// The collection of previously applied audio influencers.
+        /// </summary>
+        private List<IAudioInfluencer> previousInfluencers = new List<IAudioInfluencer>();
+
+        /// <summary>
+        /// Potential effects manipulated by an audio influencer and their key
+        /// properties.
+        /// <summary>
+        private AudioLowPassFilter lowPassFilter;
+        private float nativeLowPassCutoffFrequency;
+        private AudioHighPassFilter highPassFilter;
+        private float nativeHighPassCutoffFrequency;
+
+        private void Awake() 
+        {           
+            audioSource = gameObject.GetComponent<AudioSource>();
+            initialAudioSourceVolume = audioSource.volume;
+
+            // Get optional filters (and initial values) that the sound designer / developer 
+            // may have applied to this game object
+            lowPassFilter = gameObject.GetComponent<AudioLowPassFilter>();
+            nativeLowPassCutoffFrequency = (lowPassFilter != null) ? lowPassFilter.cutoffFrequency : NeutralHighFrequency;
+            highPassFilter = gameObject.GetComponent<AudioHighPassFilter>();
+            nativeHighPassCutoffFrequency = (highPassFilter != null) ? highPassFilter.cutoffFrequency : NeutralLowFrequency;
+
+            // Preallocate the array that will be used to collect RaycastHit structures.
+            hits = new RaycastHit[MaxObjects];
+        }
+    
+        private void Update() 
+        {
+            DateTime now = DateTime.Now;
+
+            // Audio influences are not updated every frame.
+            if ((UpdateInterval * 1000.0f) <= (now - lastUpdate).Milliseconds)
+            {
+                audioSource.volume = initialAudioSourceVolume;
+
+                // Get the audio influencers that should apply to the audio source.
+                List<IAudioInfluencer> influencers = GetInfluencers();
+                foreach (IAudioInfluencer influencer in influencers)
+                {
+                    // Apply the influencer's effect.
+                    influencer.ApplyEffect(gameObject);
+                }
+
+                // Find and remove the audio influencers that are to be removed from the audio source.
+                List<IAudioInfluencer> influencersToRemove = new List<IAudioInfluencer>();
```

Comment:
Do we want to make a new list each time?

Or clear the one we have and fill it again?

Reply:
since callers iterate over what is returned, we should not change the contents. can cause exceptions if the caller uses foreach and potentially walking off the end of the collection / missing items in the case of for loops.

Reply:
this is especially problematic if a project is using a Coroutine and the collection gets changed during a yield

Reply:
Let's tend to use async instead of coroutines... maybe we should write that up as well.

Reply:
it is not for us to decide how client applications are written, :)

Reply:
of course! lol,

but, internally we can def keep to using async over corountines